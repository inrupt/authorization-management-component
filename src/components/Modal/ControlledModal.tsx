//
// Copyright Inrupt Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to use,
// copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
// Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

import { useState, type ReactNode } from "react";
import { Modal, ModalBody, ModalFooter, ModalHeader } from "reactstrap";
import styles from "./Modal.module.scss";
import Button from "../Button/Button";

export interface Props {
  isOpen: boolean;
  onClose: () => void;
  onPrimaryAction?: () => void;
  primaryActionText?: string;
  title: string;
  confirmationImage?: string;
  onSecondaryAction?: () => void;
  secondaryActionText?: string;
  children?: ReactNode;
  variant?: string;
  isDanger?: boolean;
}

export default function ControlledModal({
  isOpen,
  onClose,
  confirmationImage,
  onPrimaryAction,
  primaryActionText,
  onSecondaryAction,
  secondaryActionText,
  title,
  children,
  variant,
  isDanger,
}: Props) {
  const [transitionedIn, setTransitionedIn] = useState<boolean>(false);
  const [transitioningOut, setTransitioningOut] = useState<boolean>(false);
  return (
    <Modal
      isOpen={isOpen}
      toggle={onClose}
      labelledBy="modalLabel"
      className={
        variant === "confirmation" ? styles["confirmation-modal"] : undefined
      }
      data-testid={
        transitionedIn && !transitioningOut ? "modal-ready" : undefined
      }
      onEnter={() => setTransitionedIn(false)}
      onOpened={() => setTransitionedIn(true)}
      onExit={() => setTransitioningOut(true)}
      onClosed={() => setTransitioningOut(false)}
    >
      <ModalHeader
        className={
          variant == "warn"
            ? styles["warn-modal-header"]
            : styles["modal-header"]
        }
      >
        <div>
          {confirmationImage && (
            <i className={confirmationImage} title="Confirm" />
          )}
          <h4 className="modal-title" id="modalLabel">
            {title}
          </h4>
        </div>
      </ModalHeader>
      <ModalBody
        className={
          variant == "warn"
            ? styles["warn-modal-content"]
            : styles["modal-content"]
        }
      >
        {children}
      </ModalBody>
      <ModalFooter
        className={
          variant == "warn"
            ? styles["warn-modal-footer"]
            : styles["modal-footer"]
        }
      >
        {secondaryActionText && (
          <Button
            type="button"
            variant="outline"
            className={`btn btn-secondary ${styles["modal-secondary-btn"]}`}
            onClick={onSecondaryAction}
          >
            {secondaryActionText}
          </Button>
        )}
        <Button
          type="button"
          variant={isDanger ? "danger" : "primary"}
          data-testid="modal-primary-action"
          className={`btn btn-primary ${styles["modal-primary-btn"]}`}
          onClick={onPrimaryAction ?? onClose}
        >
          {primaryActionText}
        </Button>
      </ModalFooter>
    </Modal>
  );
}

ControlledModal.defaultProps = {
  variant: "default",
  primaryActionText: "Save",
  onPrimaryAction: undefined,
  confirmationImage: undefined,
  onSecondaryAction: undefined,
  secondaryActionText: undefined,
  children: undefined,
};
