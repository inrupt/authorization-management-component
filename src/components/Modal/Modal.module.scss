@use "../../../scss/_variables" as *;

@import "../../../node_modules/bootstrap/scss/mixins";

.modal-content {
  padding: 0rem 2rem 1.9rem 2rem;
}

.modal-header,
.warn-modal-header {
  padding: 2.5rem 0rem 1rem 2rem;
  border-bottom: none;
  h4 {
    font-size: 1.125rem;
  }
}

.modal-footer,
.warn-modal-footer {
  border-top: none;
  padding: 0rem 2rem 2.5rem 0rem;
  gap: 1.5rem;
  button {
    margin: 0rem;
  }
}

.warn-modal-content {
  padding: 0rem 6rem 1.45rem 2.5rem;
}

.warn-modal-header {
  padding: 2rem 2.2rem 1rem 2.5rem;
  h4 {
    font-weight: 700;
  }
}

.warn-modal-footer {
  gap: 0rem;
  padding: 0rem 2.2rem 1.45rem 0rem;
  button {
    padding: 0.5rem 1.1rem 0.6rem 1.2rem;
  }
}

.close {
  cursor: pointer;
  background: none;
  border: none;
  width: max-content;
  align-self: flex-end;
  font-size: 1.25rem;
}

.confirmation-modal {
  text-align: center;

  .modal-header {
    display: flex;
    justify-content: center;
    flex-direction: column;
    padding-inline: 0rem;
  }

  .modal-footer {
    padding-inline: 0rem;
    justify-content: center;
  }

  i {
    font-size: 3rem;
  }
}

.modal-secondary-btn {
  color: $black;
}
