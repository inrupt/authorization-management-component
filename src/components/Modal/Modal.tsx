// MIT License
//
// Copyright Inrupt Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
//

import { useState } from "react";
import styles from "./Modal.module.scss";
import type { Props as ControlledModalProps } from "./ControlledModal";
import ControlledModal from "./ControlledModal";
import Button from "../Button/Button";

interface Props
  extends Omit<
    ControlledModalProps,
    "isOpen" | "onClose" | "onPrimaryAction" | "onSecondaryAction"
  > {
  modalTriggerText: string;
  variant?: string;
  onPrimaryAction?: () => void;
  testIdSuffix?: string;
  onClick?: () => void;
}

export default function Modal({
  modalTriggerText,
  variant = "default",
  testIdSuffix = "",
  onClick,
  ...props
}: Props) {
  const [isOpen, setOpen] = useState(false);
  const onClose = () => setOpen(false);

  return (
    <>
      {modalTriggerText && (
        <Button
          type="button"
          data-testid={`modal-trigger${testIdSuffix ? `-${testIdSuffix}` : ""}`}
          className={`btn btn-primary ${styles["modal-primary-btn"]}`}
          variant={variant}
          onClick={(e) => {
            e.stopPropagation();
            setOpen(true);
            if (onClick !== undefined) {
              onClick();
            }
          }}
        >
          {modalTriggerText}
        </Button>
      )}

      <ControlledModal
        {...props}
        onClose={onClose}
        onPrimaryAction={() => {
          onClose();
          props.onPrimaryAction?.();
        }}
        onSecondaryAction={onClose}
        isOpen={isOpen}
        variant={variant}
      />
    </>
  );
}
