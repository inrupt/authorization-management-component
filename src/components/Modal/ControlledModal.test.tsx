//
// Copyright Inrupt Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to use,
// copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
// Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

import { render, waitFor } from "@testing-library/react";
import * as axe from "axe-core";
import Modal from "./ControlledModal";

describe("Modal", () => {
  it.each(["danger", "default", "primary"])(
    "renders an %s Modal",
    async (variant) => {
      const { baseElement, getByText } = render(
        <Modal
          title="Deny All Access?"
          isOpen
          onClose={() => {
            /* noop */
          }}
          variant={variant}
          isDanger={variant === "danger"}
        >
          Agent X will not have access to this resource.
        </Modal>
      );

      await waitFor(() => {
        return expect(getByText("Deny All Access?")).toBeVisible();
      });

      // Accessibility tests
      HTMLCanvasElement.prototype.getContext = jest.fn();
      const results = await axe.run(baseElement);
      expect(results.violations).toHaveLength(0);

      expect(baseElement).toMatchSnapshot();
    }
  );
});
