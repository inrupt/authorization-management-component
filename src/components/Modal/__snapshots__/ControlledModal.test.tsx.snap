// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders an danger Modal 1`] = `
<body
  class="modal-open"
  style="overflow: hidden;"
>
  <div />
  <div
    style="position: relative; z-index: 1050; display: block;"
    tabindex="-1"
  >
    <div
      class=""
    >
      <div
        aria-labelledby="modalLabel"
        aria-modal="true"
        class="modal fade"
        role="dialog"
        style="display: block;"
        tabindex="-1"
      >
        <div
          class="modal-dialog"
          role="document"
        >
          <div
            class="modal-content"
          >
            <div
              class="modal-header modal-header"
            >
              <h5
                class="modal-title"
              >
                <div>
                  <h4
                    class="modal-title"
                    id="modalLabel"
                  >
                    Deny All Access?
                  </h4>
                </div>
              </h5>
            </div>
            <div
              class="modal-content modal-body"
            >
              Agent X will not have access to this resource.
            </div>
            <div
              class="modal-footer modal-footer"
            >
              <button
                class="button danger btn btn-primary modal-primary-btn"
                data-testid="modal-primary-action"
                type="button"
              >
                Save
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="modal-backdrop fade"
      />
    </div>
  </div>
</body>
`;

exports[`Modal renders an default Modal 1`] = `
<body
  class="modal-open"
  style="overflow: hidden;"
>
  <div />
  <div
    style="position: relative; z-index: 1050; display: block;"
    tabindex="-1"
  >
    <div
      class=""
    >
      <div
        aria-labelledby="modalLabel"
        aria-modal="true"
        class="modal fade"
        role="dialog"
        style="display: block;"
        tabindex="-1"
      >
        <div
          class="modal-dialog"
          role="document"
        >
          <div
            class="modal-content"
          >
            <div
              class="modal-header modal-header"
            >
              <h5
                class="modal-title"
              >
                <div>
                  <h4
                    class="modal-title"
                    id="modalLabel"
                  >
                    Deny All Access?
                  </h4>
                </div>
              </h5>
            </div>
            <div
              class="modal-content modal-body"
            >
              Agent X will not have access to this resource.
            </div>
            <div
              class="modal-footer modal-footer"
            >
              <button
                class="button primary btn btn-primary modal-primary-btn"
                data-testid="modal-primary-action"
                type="button"
              >
                Save
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="modal-backdrop fade"
      />
    </div>
  </div>
</body>
`;

exports[`Modal renders an primary Modal 1`] = `
<body
  class="modal-open"
  style="overflow: hidden;"
>
  <div />
  <div
    style="position: relative; z-index: 1050; display: block;"
    tabindex="-1"
  >
    <div
      class=""
    >
      <div
        aria-labelledby="modalLabel"
        aria-modal="true"
        class="modal fade"
        role="dialog"
        style="display: block;"
        tabindex="-1"
      >
        <div
          class="modal-dialog"
          role="document"
        >
          <div
            class="modal-content"
          >
            <div
              class="modal-header modal-header"
            >
              <h5
                class="modal-title"
              >
                <div>
                  <h4
                    class="modal-title"
                    id="modalLabel"
                  >
                    Deny All Access?
                  </h4>
                </div>
              </h5>
            </div>
            <div
              class="modal-content modal-body"
            >
              Agent X will not have access to this resource.
            </div>
            <div
              class="modal-footer modal-footer"
            >
              <button
                class="button primary btn btn-primary modal-primary-btn"
                data-testid="modal-primary-action"
                type="button"
              >
                Save
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="modal-backdrop fade"
      />
    </div>
  </div>
</body>
`;
